bokeh/plotting/contour.py

```diff
@@ -17,23 +17,73 @@
 # Imports
 #-----------------------------------------------------------------------------
 
+# Standard library imports
+from typing import Any, Dict, Optional, Sequence, Tuple, Union
+
 # External imports
 import numpy as np
 
+# Bokeh imports
+from ..models.renderers import ContourRenderer
+
 #-----------------------------------------------------------------------------
 # Globals and constants
 #-----------------------------------------------------------------------------
 
 __all__ = (
     'contour_data_dicts',
     'contour_coords',
+    'from_contour',
 )
 
 #-----------------------------------------------------------------------------
 # General API
 #-----------------------------------------------------------------------------
 
-def contour_coords(x, y, z, levels, want_fill, want_line):
+def from_contour(
+    x: Optional[np.ndarray],
```

Comment:
`np.ndarray` should be replaced with `npt.NDArray[float]`, as mypy doesn't like the former (`import numpy.typing as npt`). We prefer `x: T | None` instead of `x: Optional[T]`, because such arguments are rarely truly optional, especially if there's no default value provided.

Reply:
I've been using `Optional[T]` rather than `T | None` in other projects for backwards compatibility. It didn't even occur to me that the pipe would be OK here.

`x` and `y` can be truly optional eventually as they are optional in contourpy.

Reply:
Pipe is OK because we use `from __future__ import annotations`, which makes annotations lazy, and thus we don't care whether Python supports `|` on types, only mypy. There is a caveat though. Type aliases have to use `Union[]`, as those are always evaluated eagerly (one of joys of static typing in Python).

Reply:
Thanks for the clarification.